### DISCUSSION TITLE
Storing ActionData in the Model

### DISCUSSION THREAD

**Discussion by pyrex41** [dillonkearns/elm-pages#454] [General]
1) Is it conceptually possible to store actionData in the model? The actionData is already available to `view`, but if another action is triggered after, that actionData is lost, replaced with the new actionData (I think).

2) Is this a good idea? Or is there a better approach.

**Comment by dillonkearns** [comment]
`ActionData` is intentionally passed in as a value to be consumed declaratively wherever you need it, and not to be stored in the `Model`. If you do so then you risk getting out of sync, so I would recommend against it to avoid bugs and complexity.

The idea in `elm-pages` is to derive state as much as possible from the framework state (`Data`, `ActionData`, pending form requests, etc.). These pieces of state are managed by the framework for you and passed along where you would need them, and the general advice in Elm is to derive state unless for performance reasons you need to do otherwise, because the tradeoff is that you introduce more surface area for things to get out of sync and you lose your clear source of truth. This is even more true with `ActionData` because the elm-pages framework doesn't expose low-level ways to track when the `ActionData` state has changed and listen to updates to the data, so it would end up getting hacky and probably cause issues.

Is your use case getting the logged in user over to `Shared.view`, or something else?

**Reply by pyrex41** [reply]
Similar, I was looking to update the view with a Success / Fail message based on something uploaded via a form. Which wasn't working via update/model, since the model isn't updated on receiving the action response. But turns out it's easy to do since view has `action` as an arg -- I just started wondering, what if I need to know about this response later. How can I keep track of what happened.

**Reply by dillonkearns** [reply]
I see, I don't think I understand the specifics of it well enough to know what you would want to do with the data in the future (rather than just the latest `ActionData`). There is actually this helper:

```elm
RouteBuilder.withOnAction : (action -> msg) -> StatefulRoute routeParams data action model msg -> StatefulRoute routeParams data action model msg
```

So you can do a lot with that by subscribing to those events as well. Be careful about stale data though, that is definitely worth avoiding.

**Reply by pyrex41** [reply]
Thank you, that helper would handle the scenario I was considering quite nicely. Elm-pages has been great, but most of the pitfalls I've faced have been caused my not fully understanding the framework and reaching instead to force a more vanilla-elm approach that ends up being clunky. This is really helpful to clarify how/when to use `action` / `ActionData`
